### PR TITLE
feat(updated): Add silent install capability

### PR DESCRIPTION
**Pull request by jaredburkeen**
## What's New?

-  Added automatic download and silent install of available updates
- Extended appcast callback to allow for silent installs (upgrade)

### CODE REVIEW

**Comment by jaredburkeen** [comment]
@loberlander This is ready for your 👀

File: src/updatechecker.cpp [hunk 1]
```diff
@@ -254,32 +289,65 @@ void UpdateChecker::PerformUpdateCheck(bool manual)
...
-            UI::NotifyNoUpdates(ShouldAutomaticallyInstall());
-            return;
+                if (Settings::HasDSAPubKeyPem())
+                {
+                    SignatureVerifier::VerifyDSASHA1SignatureValid(sink.GetFilePath(), appcast.DsaSignature);
+                }
+                else
+                {
+                    // backward compatibility - accept as is, but complain about it
+                    LogError("Using unsigned updates!");
```
Comment (loberlander) [inline src/updatechecker.cpp:326]: Why do we allow this? This seems like a back door to bypass checking in the future.

File: src/updatechecker.cpp [hunk 2]
```diff
@@ -254,32 +289,65 @@ void UpdateChecker::PerformUpdateCheck(bool manual)
...
 
-        // Check if the user opted to ignore this particular version.
-        if ( ShouldSkipUpdate(appcast) )
-        {
-            UI::NotifyNoUpdates(ShouldAutomaticallyInstall());
-            return;
+                if (Settings::HasDSAPubKeyPem())
+                {
+                    SignatureVerifier::VerifyDSASHA1SignatureValid(sink.GetFilePath(), appcast.DsaSignature);
+                }
```
Comment (loberlander) [inline src/updatechecker.cpp:322]: Shouldn't there be an action of some kind if the signature check fails to block the update?

File: src/updatechecker.cpp [hunk 3]
```diff
@@ -254,32 +289,65 @@ void UpdateChecker::PerformUpdateCheck(bool manual)
...
+            // Check if our version is out of date.
+            if (!appcast.IsValid() || CompareVersions(currentVersion, appcast.Version) >= 0)
+            {
+                // The same or newer version is already installed.
+                return;
+            }
 
-        const std::string currentVersion =
-                WideToAnsi(Settings::GetAppBuildVersion());
+            if (appcast.DownloadURL != "")
```
Comment (loberlander) [inline src/updatechecker.cpp:310]: Would `!appcast.DownloadURL.empty()` be better here?

File: src/updatechecker.cpp [hunk 4]
```diff
@@ -219,6 +221,39 @@ UpdateChecker::UpdateChecker(): Thread("WinSparkle updates check")
...
+
+    if (!CreateProcess(NULL, const_cast<WCHAR*>(fullCommandLine.c_str()), NULL, NULL, TRUE, DETACHED_PROCESS, NULL, NULL, &Startup, lpProcessInformation))
+    {
+        return false;
+    }
+
+    if (lpProcessInformation == &processInfo)
+    {
+        CloseHandle(processInfo.hProcess);
+        CloseHandle(processInfo.hThread);
```
Comment (loberlander) [inline src/updatechecker.cpp:251]: Is the caller expected to do CloseHandle() on `lpProcessInformation.hProcess` and `lpProcessInformation.hThread` after `true` is returned? Isn't it too risky to allow/require the caller to do that? Is that how it is going to be used? Why does the caller need to have the Process Info after completion? Or the other way around, if the caller does NOT pass in a Process Info pointer, why can't we just return with an error and not worry about the local Process Info?

File: src/updatechecker.cpp [hunk 5]
```diff
@@ -219,6 +221,39 @@ UpdateChecker::UpdateChecker(): Thread("WinSparkle updates check")
...
+    PROCESS_INFORMATION     processInfo;
+
+    if (filePath == L"")
+    {
+        return false;
+    }
+
+    GetStartupInfo(&Startup);
+    Startup.wShowWindow = SW_SHOWNORMAL;
+    if (lpProcessInformation == NULL)
```
Comment (loberlander) [inline src/updatechecker.cpp:236]: Maybe a comment here would be nice about why a local Process Info is desirable to use when there is no caller provided Process Info.

File: src/updatedownloader.cpp [hunk 6]
```diff
@@ -44,10 +41,7 @@ namespace winsparkle
                                   helpers
  *--------------------------------------------------------------------------*/
 
-namespace
-{
-
-std::wstring GetUniqueTempDirectoryPrefix()
+std::wstring winsparkle::GetUniqueTempDirectoryPrefix()
```
Comment (loberlander) [inline src/updatedownloader.cpp:44]: Is name space designation `winsparkle::` really needed here when already inside the namespace?

File: src/updatedownloader.cpp [hunk 7]
```diff
@@ -58,7 +52,7 @@ std::wstring GetUniqueTempDirectoryPrefix()
     return dir;
 }
 
-std::wstring CreateUniqueTempDirectory()
+std::wstring winsparkle::CreateUniqueTempDirectory()
```
Comment (loberlander) [inline src/updatedownloader.cpp:55]: Is name space designation `winsparkle::` really needed here when already inside the namespace?

File: src/updatedownloader.cpp [hunk 8]
```diff
@@ -173,7 +102,7 @@ void UpdateDownloader::Run()
 
     try
     {
-      const std::wstring tmpdir = CreateUniqueTempDirectory();
+      const std::wstring tmpdir = winsparkle::CreateUniqueTempDirectory();
```
Comment (loberlander) [inline src/updatedownloader.cpp:105]: Is name space designation `winsparkle::` really needed here when already inside the namespace?

File: src/updatedownloader.cpp [hunk 9]
```diff
@@ -222,7 +151,7 @@ void UpdateDownloader::CleanLeftovers()
     // malicious users from forcing us into deleting arbitrary directories:
     try
     {
-        if (tmpdir.find(GetUniqueTempDirectoryPrefix()) != 0)
+        if (tmpdir.find(winsparkle::GetUniqueTempDirectoryPrefix()) != 0)
```
Comment (loberlander) [inline src/updatedownloader.cpp:154]: Is name space designation `winsparkle::` really needed here when already inside the namespace?

File: src/updatedownloader.h [hunk 10]
```diff
@@ -28,42 +28,111 @@
...
-    and shows information about available update to the user.
- */
-class UpdateDownloader : public Thread
-{
-public:
-    /// Creates updater thread.
-    UpdateDownloader(const Appcast& appcast);
+    // helper functions
+    std::wstring GetUniqueTempDirectoryPrefix();
+    std::wstring CreateUniqueTempDirectory();
```
Comment (loberlander) [inline src/updatedownloader.h:41]: Are these used externally? Do they need to be exposed in the header file?

File: src/updatedownloader.h [hunk 11]
```diff
@@ -28,42 +28,111 @@
...
+        // Thread methods:
+        virtual void Run();
+        virtual bool IsJoinable() const { return true; }
+
+    private:
+        Appcast m_appcast;
+    };
+
+
+    struct UpdateDownloadSink : public IDownloadSink
```
Comment (loberlander) [inline src/updatedownloader.h:74]: Why is this implementation (definition) moving into the header file? Shouldn't it stay in the cpp file and just have a declaration here in the header file if it needs to be exposed?

File: src/updatechecker.cpp (hunk 2 above)
Comment (loberlander) [reply src/updatechecker.cpp:316]: With silent updates it is probably more likely that a malicious install would go undetected. The SHA value is more of an integrity test than a security feature as a malicious actor could fame that as well. So I am not certain how far we really need to go with this.
This topic is probably of interest to the security team so they are probably better suited to answer.

File: src/updatechecker.cpp (hunk 4 above)
Comment (loberlander) [reply src/updatechecker.cpp]: I would tailor it for your exact use case, i.e. either make lpProcessInformation a mandatory not NULL parameter OR I would remove it altogether.

Comment (jaredburkeen) [reply src/updatechecker.cpp]: Yes, I removed that parameter altogether.

File: src/updatechecker.cpp (hunk 2 above)
Comment (jaredburkeen) [reply src/updatechecker.cpp:316]: Good idea, I will solicit the security team for feedback.
@andrewkeller I'm curious if you added any additional security when you added silent updates for Mac? For Windows we're downloading the update installer using the client api.

**Comment by jaredburkeen** [comment]
@loberlander This is ready for further review. Waiting on some feedback from security so there could be additional tweaks there.

**Pull request merged**